gradio/blocks.py

```diff
@@ -2684,17 +2682,42 @@ def reverse(text):
                     "In order to use `mcp_server=True`, you must install gradio with the `mcp` extra. Please install it with `pip install gradio[mcp]`"
                 ) from e
             try:
-                self.mcp_server_obj = gradio.mcp.GradioMCPServer(self)
-                self.mcp_server_obj.launch_mcp_on_sse(
-                    self.server_app, mcp_subpath, self.root_path
-                )
+                self.mcp_server_obj = gradio.mcp.GradioMCPServer(self, self.root_path)
                 self.mcp_server = True
+                user_lifespan = None
```

Comment:
The StreamableHTTPSessionManager's run method must be called when the app [starts](https://github.com/modelcontextprotocol/python-sdk/blob/05b7156ea8a34d8476a7cfbef5f754e22ab6c697/examples/servers/simple-streamablehttp/mcp_simple_streamablehttp/server.py#L149)

FastAPI will not run the lifespan of mounted subapplications so we need to pass the mcp lifespan all the way back up to `App.create_app`. 

This refactor switches the order of creating the app and the mcp server so that the mcp server and lifespan are defined before the gradio server.

There is also some logic here to ensure the mcp server lifespan works well with a user defined lifespan.